README.md
# Salte Filter
[![Slack Status][slack-image]][slack-url]
[![NPM Version][npm-version-image]][npm-url]
[![NPM Downloads][npm-downloads-image]][npm-url]
[![Travis][travis-ci-image]][travis-ci-url]
[![Coveralls][coveralls-image]][coveralls-url]

[![Commitizen friendly][commitizen-image]][commitizen-url]
[![semantic-release][semantic-release-image]][semantic-release-url]

Registers filter components allowing for a single input and output!

## Install

You can install this package either with `npm` or with `bower`.

## npm

```sh
$ npm install salte-filter
```
Then add a `<script>` to your index.html:
```html
<script src="/node_modules/salte-filter/salte-filter.js"></script>
```
Or `require('salte-filter')` from your code.

## bower

```sh
$ bower install salte-io/salte-filter
```
Then add a `<script>` to your index.html:
```html
<script src="/bower_components/salte-filter/salte-filter.js"></script>
```

## Example

[Live Demo](https://jsbin.com/guyoyelupe/edit?html,js,output)

## Documentation

`salteFilter` must be required into your component in order to be used

### `(add/remove)Filter`
* **`filter`:** A filter to be called anytime the filter function is called

```javascript
const component = {
  require: {
    parent: '^salteFilter'
  },
  controller: () => {
    const view = this;

    view.filter = (item) => {
      // ... returns the item if the function returns truthy
    };

    view.$onInit = () => {
      // This will add the filter
      view.parent.addFilter(view.filter);
      // This will remove the filter
      view.parent.removeFilter(view.filter);
    };
  }
}
```

### `(add/remove)FilterListener`
* **`listener`:** A listener to be called anytime the filter function is called

```javascript
const component = {
  require: {
    parent: '^salteFilter'
  },
  controller: () => {
    const view = this;

    view.onFilter = (filteredData) => {
      // ...
    };

    view.$onInit = () => {
      // This will add the listener
      view.parent.addFilterListener(view.onFilter);
      // This will remove the listener
      view.parent.removeFilterListener(view.onFilter);
    };
  }
}
```

### `filter`
Runs all of the filters, useful for triggering a filter when an input value is changed

## License

The MIT License

Copyright (c) 2012-2016 Salte. https://salte.io

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

[slack-image]: https://salte-slack.herokuapp.com/badge.svg
[slack-url]: https://salte-slack.herokuapp.com/

[npm-version-image]: http://img.shields.io/npm/v/salte-filter.svg?style=flat
[npm-downloads-image]: http://img.shields.io/npm/dm/salte-filter.svg?style=flat
[npm-url]: https://npmjs.org/package/salte-filter

[travis-ci-image]: https://img.shields.io/travis/salte-io/salte-filter.svg?style=flat
[travis-ci-url]: https://travis-ci.org/salte-io/salte-filter

[coveralls-image]: https://img.shields.io/coveralls/salte-io/salte-filter/master.svg
[coveralls-url]: https://coveralls.io/github/salte-io/salte-filter

[commitizen-image]: https://img.shields.io/badge/commitizen-friendly-brightgreen.svg
[commitizen-url]: http://commitizen.github.io/cz-cli/

[semantic-release-url]: https://github.com/semantic-release/semantic-release
[semantic-release-image]: https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg


CHANGELOG.md
# CHANGELOG

The changelog is automatically updated using [semantic-release](https://github.com/semantic-release/semantic-release).
You can see it on the [releases page](https://github.com/salte-io/salte-filter/releases).


tests/salte-filter/salte-filter.spec.js
import salteFilter from '../../src/salte-filter.module.js';

describe('component(salte-filter)', () => {
  let scope, component, controller;

  beforeEach(angular.mock.module(salteFilter));
  beforeEach(angular.mock.inject(($compile, $rootScope) => {
    scope = $rootScope.$new();
    scope.data = [{ business: 'Salte' }, { business: 'Google' }];
    component = $compile('<salte-filter data="data" filtered-data="filteredData"></salte-filter>')(scope);
    controller = component.controller('salteFilter');
    scope.$digest();
  }));

  describe('function(filter)', () => {
    it('should return a dereferenced version of the original data if no filters are defined', () => {
      expect(scope.filteredData).toEqual(scope.data);
      scope.filteredData.push({ business: 'Valve' });
      expect(scope.filteredData).not.toEqual(scope.data);
    });

    it('should filter when a filter is added or removed', () => {
      function filter(data) {
        return data.business === 'Salte';
      }

      expect(scope.filteredData).toEqual(scope.data);
      controller.addFilter(filter);
      scope.$digest();
      expect(scope.filteredData).toEqual([{ business: 'Salte' }]);
      controller.removeFilter(filter);
      scope.$digest();
      expect(scope.filteredData).toEqual(scope.data);
    });

    it('should filter when the input data is modified', () => {
      expect(scope.filteredData).toEqual(scope.data);
      scope.data = [{ restaurant: 'Taco Bell' }];
      scope.$digest();
      expect(scope.filteredData).toEqual(scope.data);
    });

    it('should not filter when an individual element is pushed or popped from the input data', () => {
      expect(scope.filteredData).toEqual(scope.data);
      scope.data.push({ business: 'Twitch' });
      scope.$digest();
      expect(scope.filteredData).toEqual([{ business: 'Salte' }, { business: 'Google' }]);
    });

    it('should allow the add a filter listener', () => {
      const listener = sinon.spy();
      controller.addFilterListener(listener);
      controller.filter();
      expect(listener.callCount).toEqual(1);
      controller.removeFilterListener(listener);
      controller.filter();
      expect(listener.callCount).toEqual(1);
    });
  });
});


karma.conf.js
const webpackConfig = require('./webpack.test.config.js');
webpackConfig.module.rules.push({
  enforce: 'pre',
  test: /\.js$/,
  exclude: /tests|node_modules/,
  loader: 'isparta'
});

module.exports = function(config) {
  config.set({
    basePath: '',

    frameworks: [
      'jasmine',
      'sinon'
    ],

    files: [
      'tests/index.js'
    ],

    preprocessors: {
      'tests/index.js': ['webpack', 'sourcemap']
    },

    webpack: webpackConfig,

    webpackMiddleware: {
      noInfo: true
    },

    reporters: ['mocha', 'coverage'],

    coverageReporter: {
      dir: 'coverage',
      reporters: [
        { type: 'text' },
        { type: 'lcovonly', subdir: '.', file: 'lcov.info' }
      ]
    },

    port: 9876,

    colors: true,

    logLevel: config.LOG_INFO,

    autoWatch: true,

    browsers: ['PhantomJS'],

    browserNoActivityTimeout: 120000,

    singleRun: false
  });
};


src/salte-filter.controller.js
export default class SalteFilter {
  constructor() {
    this.filters = [];
    this.listeners = [];
  }

  $onChanges(changes) {
    if (changes.data) {
      this.filter();
    }
  }

  filter() {
    this.filteredData = [];
    if (this.data) {
      this.data.forEach((data) => {
        let valid = true;
        this.filters.forEach((filter) => {
          if (!filter(data)) {
            valid = false;
          }
        });
        if (valid) {
          this.filteredData.push(data);
        }
      });
    }
    this.listeners.forEach((listener) => listener(this.filteredData));
  }

  addFilter(filter) {
    if (this.filters.indexOf(filter) === -1) {
      this.filters.push(filter);
      this.filter();
    }
  }

  addFilterListener(listener) {
    if (!listener) return;
    this.listeners.push(listener);
  }

  removeFilterListener(listener) {
    const index = this.listeners.indexOf(listener);
    if (index !== -1) {
      this.listeners.splice(index, 1);
    }
  }

  removeFilter(filter) {
    const index = this.filters.indexOf(filter);
    if (index !== -1) {
      this.filters.splice(index, 1);
      this.filter();
    }
  }
}


karma.ci.conf.js
const webpackConfig = require('./webpack.test.config.js');

module.exports = function(config) {
  const customLaunchers = {
    Chrome: {
      base: 'SauceLabs',
      browserName: 'chrome'
    },
    Firefox: {
      base: 'SauceLabs',
      browserName: 'firefox'
    },
    Edge: {
      base: 'SauceLabs',
      browserName: 'microsoftedge'
    },
    InternetExplorer11: {
      base: 'SauceLabs',
      browserName: 'internet explorer',
      version: '11'
    },
    InternetExplorer10: {
      base: 'SauceLabs',
      browserName: 'internet explorer',
      version: '10'
    },
    Safari9: {
      base: 'SauceLabs',
      browserName: 'safari',
      version: '9'
    },
    Safari8: {
      base: 'SauceLabs',
      browserName: 'safari',
      version: '8'
    },
    Safari7: {
      base: 'SauceLabs',
      browserName: 'safari',
      version: '7'
    }
  };

  config.set({
    basePath: '',

    frameworks: [
      'jasmine',
      'sinon'
    ],

    files: [
      'tests/index.js'
    ],

    preprocessors: {
      'tests/index.js': ['webpack', 'sourcemap']
    },

    webpack: webpackConfig,

    webpackMiddleware: {
      noInfo: true
    },

    reporters: ['spec', 'saucelabs'],

    port: 9876,

    colors: true,

    logLevel: config.LOG_INFO,

    sauceLabs: {
      testName: 'salte-io/salte-filter',
      tunnelIdentifier: process.env.TRAVIS_JOB_NUMBER,
      startConnect: false
    },

    customLaunchers: customLaunchers,
    browsers: Object.keys(customLaunchers),

    captureTimeout: 0,
    browserNoActivityTimeout: 120000,

    singleRun: true
  });
};
